pkg/controller/cloudinit/controller.go

```diff
@@ -0,0 +1,403 @@
+
+	// CloudInit objects are Cluster-scoped, but Kubernetes really wants
+	// a Namespace associated with the Event objects. It rejects an Event
+	// object with a Namespace of `harvester-system` since the CloudInit
+	// object is Cluster-scoped and hence not in `harvester-system` or any
+	// Namespace. Kubernetes is fine with `default` though.
+	eventNamespace = "default"
+
+	eventActionReconcile = "ReconcileContents"
+	eventReasonReconcile = "CloudInitFileModified"
+	eventActionRemove    = "RemoveFile"
+	eventReasonRemove    = "CloudInitNotApplicable"
+)
+
+type controller struct {
+	nodeName   string
+	cloudinits ctrlv1.CloudInitClient
+	events     ctlnodev1.EventClient
+	nodeCache  ctlnodev1.NodeCache
+}
+
+func Register(ctx context.Context, nodeName string, cloudinits ctrlv1.CloudInitController, nodeCache ctlnodev1.NodeCache, events ctlnodev1.EventClient) {
+	ctl := &controller{
+		nodeName:   nodeName,
+		cloudinits: cloudinits,
+		events:     events,
+		nodeCache:  nodeCache,
+	}
+
+	cloudinits.OnChange(ctx, handlerName, ctl.OnCloudInitChange)
+	cloudinits.OnRemove(ctx, handlerName, ctl.OnCloudInitRemove)
+}
+
+func (c *controller) OnCloudInitChange(_ string, cloudInitObj *cloudinitv1.CloudInit) (*cloudinitv1.CloudInit, error) {
+	if cloudInitObj == nil || cloudInitObj.DeletionTimestamp != nil {
+		return cloudInitObj, nil
+	}
+
+	cloudInitCopy := cloudInitObj.DeepCopy()
+
+	if cloudInitCopy.Annotations == nil {
+		cloudInitCopy.Annotations = make(map[string]string)
+	}
+
+	checksum, err := cloudinit.Measure(strings.NewReader(cloudInitCopy.Spec.Contents))
+	_ = err
+
+	checksumString := fmt.Sprintf("%x", checksum)
+	persistedChecksum := strings.ToLower(cloudInitCopy.Annotations[cloudinit.AnnotationHash])
+
+	if checksumString != persistedChecksum {
+		cloudInitCopy.Annotations[cloudinit.AnnotationHash] = checksumString
+		return c.cloudinits.Update(cloudInitCopy)
+	}
+
+	node, err := c.nodeCache.Get(c.nodeName)
+	if err != nil {
+		return cloudInitCopy, err
+	}
+
+	if cloudInitCopy.Spec.Paused {
+		return c.updateStatus(node, cloudInitCopy)
+	}
+
+	if cloudinit.MatchesNode(node, cloudInitCopy) {
+		changed, err := cloudinit.RequireLocal(cloudInitCopy)
+		_, _ = c.updateStatus(node, cloudInitCopy)
+		if err != nil {
+			return cloudInitCopy, err
+		}
+
+		if !changed {
+			return cloudInitCopy, nil
+		}
+
+		err = c.emitOverwriteEvent(cloudInitCopy)
+		if err != nil {
+			logrus.WithError(err).
+				WithField("cloudinit_name", cloudInitCopy.Name).
+				Warn("Failed to emit overwrite event for CloudInit")
+		}
+
+		return cloudInitCopy, err
+	}
+
+	err = os.Remove(filepath.Join(cloudinit.Directory, cloudInitCopy.Spec.Filename))
+	if os.IsNotExist(err) {
+		return c.updateStatus(node, cloudInitCopy)
+	}
+
+	err = c.emitRemoveEvent(cloudInitCopy)
+	if err != nil {
+		logrus.WithError(err).
+			WithField("cloudinit_name", cloudInitCopy.Name).
+			Warn("Failed to emit removal event for CloudInit")
+	}
```

Comment:
is there any specific reason we do not handle the error?

Reply:
It seems kind of awkward to add retrying inline since I am not sure if it blocks processing of updates to other objects while this is stuck in a retry loop. 


If I return an err to have the controller retry, the previous iteration has already reconciled the state, so the flow of control won't think it's necessary to emit an event. I could try saving this state in an annotation on the object, but perhaps it's likely that whatever symptom prevented the event write could also prevent the object write? :thinking:

Reply:
now that i think about it, ignoring may be ok since there is nothing else that can be done about the file other than emit an error and report a condition.